File: src/slash.js
const { SlashCommandBuilder, PermissionFlagsBits, ChannelType, Collection } = require('discord.js');
const database = require('./database');
const help = require('./commands/help');
const setup = require('./commands/setup');
const ping = require('./commands/ping');

const EPHEMERAL = 64;
const user = (builder, required = false) => builder.addUserOption((option) => option.setName('kullanici').setDescription('Hedef kullanıcı').setRequired(required));
const reason = (builder) => builder.addStringOption((option) => option.setName('sebep').setDescription('İşlem sebebi').setMaxLength(500));

const definitions = [
  new SlashCommandBuilder().setName('yardim').setDescription('Etkileşimli yardım menüsünü açar.'),
  new SlashCommandBuilder().setName('kurulum').setDescription('Gelişmiş sunucu kurulum panelini açar.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild),
  new SlashCommandBuilder().setName('ping').setDescription('Botun bağlantı ve sistem durumunu gösterir.'),
  new SlashCommandBuilder().setName('sunucu').setDescription('Sunucu bilgilerini gösterir.'),
  user(new SlashCommandBuilder().setName('kullanici').setDescription('Kullanıcı bilgilerini gösterir.')),
  user(new SlashCommandBuilder().setName('avatar').setDescription('Kullanıcının avatarını gösterir.')),
  new SlashCommandBuilder().setName('oyunlar').setDescription('Kanal oyunlarının durumunu gösterir.'),
  new SlashCommandBuilder().setName('tdk').setDescription('TDK sözlükte ayrıntılı arama yapar.').addStringOption((o) => o.setName('kelime').setDescription('Aranacak kelime').setRequired(true).setMaxLength(80)),
  new SlashCommandBuilder().setName('webara').setDescription('Web araması yapar.').addStringOption((o) => o.setName('sorgu').setDescription('Arama sorgusu').setRequired(true).setMaxLength(500)),
  new SlashCommandBuilder().setName('zar').setDescription('Zar atar.').addStringOption((o) => o.setName('zar').setDescription('Örnek: 2d20')),
  new SlashCommandBuilder().setName('yazitura').setDescription('Yazı tura atar.'),
  new SlashCommandBuilder().setName('sekiztop').setDescription('Sihirli küreye soru sorar.').addStringOption((o) => o.setName('soru').setDescription('Sorun').setRequired(true).setMaxLength(500)),
  new SlashCommandBuilder().setName('prefix').setDescription('Komut ön ekini değiştirir.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild).addStringOption((o) => o.setName('deger').setDescription('Yeni prefix').setRequired(true).setMaxLength(5)),
  new SlashCommandBuilder().setName('ticketsetup').setDescription('Ticket sistemini tek adımda kurar.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
    .addChannelOption((o) => o.setName('kategori').setDescription('Ticket kanallarının açılacağı kategori').setRequired(true).addChannelTypes(ChannelType.GuildCategory))
    .addChannelOption((o) => o.setName('panel_kanali').setDescription('Ticket panelinin gönderileceği metin kanalı').setRequired(true).addChannelTypes(ChannelType.GuildText))
    .addChannelOption((o) => o.setName('log_kanali').setDescription('Ticket loglarının gönderileceği metin kanalı').addChannelTypes(ChannelType.GuildText))
    .addRoleOption((o) => o.setName('destek_rolu').setDescription('Ticket yetkili/destek rolü')),
  new SlashCommandBuilder().setName('ticket').setDescription('Açık ticket kanalını yönetir.')
    .addSubcommand((s) => s.setName('ekle').setDescription('Ticket kanalına kullanıcı ekler.').addUserOption((o) => o.setName('kullanici').setDescription('Eklenecek kullanıcı').setRequired(true)))
    .addSubcommand((s) => s.setName('cikar').setDescription('Ticket kanalından kullanıcı çıkarır.').addUserOption((o) => o.setName('kullanici').setDescription('Çıkarılacak kullanıcı').setRequired(true)))
    .addSubcommand((s) => s.setName('adlandir').setDescription('Ticket kanalının adını değiştirir.').addStringOption((o) => o.setName('ad').setDescription('Yeni kanal adı').setRequired(true).setMaxLength(80))),
  new SlashCommandBuilder().setName('cekilis').setDescription('Gelişmiş çekiliş oluşturur.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
    .addStringOption((o) => o.setName('sure').setDescription('10m, 2h, 3d').setRequired(true))
    .addIntegerOption((o) => o.setName('kazanan').setDescription('Kazanan sayısı').setRequired(true).setMinValue(1).setMaxValue(20))
    .addStringOption((o) => o.setName('odul').setDescription('Çekiliş ödülü').setRequired(true).setMaxLength(1000)),
  new SlashCommandBuilder().setName('cekilisyonet').setDescription('Çekilişi bitirir veya yeniden çeker.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
    .addSubcommand((s) => s.setName('bitir').setDescription('Aktif çekilişi erken bitirir.').addIntegerOption((o) => o.setName('id').setDescription('Çekiliş ID').setRequired(true).setMinValue(1)))
    .addSubcommand((s) => s.setName('yeniden').setDescription('Kazananları yeniden çeker.').addIntegerOption((o) => o.setName('id').setDescription('Çekiliş ID').setRequired(true).setMinValue(1)).addIntegerOption((o) => o.setName('kazanan').setDescription('Yeni kazanan sayısı').setMinValue(1).setMaxValue(20))),
  reason(user(new SlashCommandBuilder().setName('ban').setDescription('Kullanıcıyı onayla yasaklar.'), true))
    .addIntegerOption((o) => o.setName('mesaj_sil').setDescription('Silinecek geçmiş mesaj günü').setMinValue(0).setMaxValue(7)),
  reason(user(new SlashCommandBuilder().setName('kick').setDescription('Kullanıcıyı onayla sunucudan atar.'), true)),
  new SlashCommandBuilder().setName('unban').setDescription('Kullanıcının yasağını onayla kaldırır.').setDefaultMemberPermissions(PermissionFlagsBits.BanMembers)
    .addStringOption((o) => o.setName('kullanici_id').setDescription('Yasaklı kullanıcının ID’si').setRequired(true).setMinLength(17).setMaxLength(20))
    .addStringOption((o) => o.setName('sebep').setDescription('İşlem sebebi').setMaxLength(500)),
  new SlashCommandBuilder().setName('cases').setDescription('Moderasyon vaka geçmişini görüntüler.').setDefaultMemberPermissions(PermissionFlagsBits.ModerateMembers)
    .addUserOption((o) => o.setName('kullanici').setDescription('Belirli kullanıcının geçmişi')),
  new SlashCommandBuilder().setName('modconfig').setDescription('Moderasyon otomasyonunu ve itiraz kanalını ayarlar.').setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild)
    .addSubcommand((s) => s.setName('uyari').setDescription('Otomatik uyarı cezasını ayarla.').addIntegerOption((o) => o.setName('esik').setDescription('Aktif uyarı eşiği').setRequired(true).setMinValue(1).setMaxValue(10)).addStringOption((o) => o.setName('sure').setDescription('10m, 1h, 2d').setRequired(true)))
    .addSubcommand((s) => s.setName('itiraz').setDescription('İtiraz kanalını ayarla veya kapat.').addChannelOption((o) => o.setName('kanal').setDescription('İtiraz kanalı; boş bırakırsan kapatılır'))),
  new SlashCommandBuilder().setName('itiraz').setDescription('Yetkili ekibe gizli moderasyon itirazı gönderir.').addStringOption((o) => o.setName('metin').setDescription('Ayrıntılı itiraz metni').setRequired(true).setMinLength(20).setMaxLength(1500)),
  user(new SlashCommandBuilder().setName('mute').setDescription('Kullanıcıyı onayla susturur.'), true)
    .addStringOption((o) => o.setName('sure').setDescription('10m, 1h, 2d').setRequired(true))
    .addStringOption((o) => o.setName('sebep').setDescription('İşlem sebebi').setMaxLength(500)),
  reason(user(new SlashCommandBuilder().setName('unmute').setDescription('Susturmayı onayla kaldırır.'), true)),
  reason(user(new SlashCommandBuilder().setName('warn').setDescription('Kullanıcıya onayla uyarı verir.'), true)),
  user(new SlashCommandBuilder().setName('uyarilar').setDescription('Aktif uyarıları gösterir.')),
  user(new SlashCommandBuilder().setName('uyaritemizle').setDescription('Aktif uyarıları onayla temizler.').setDefaultMemberPermissions(PermissionFlagsBits.ModerateMembers), true),
  new SlashCommandBuilder().setName('kilit').setDescription('Kanal kilidini yönetir.').setDefaultMemberPermissions(PermissionFlagsBits.ManageChannels)
    .addStringOption((o) => o.setName('islem').setDescription('Kanalı kilitle veya aç').setRequired(true).addChoices({ name: 'Kilitle', value: 'kilitle' }, { name: 'Kilidi Aç', value: 'aç' })),
  new SlashCommandBuilder().setName('yavasmod').setDescription('Kanal yavaş modunu ayarlar.').setDefaultMemberPermissions(PermissionFlagsBits.ManageChannels)
    .addStringOption((o) => o.setName('sure').setDescription('0, 10s, 5m, 1h').setRequired(true)),
  new SlashCommandBuilder().setName('temizle').setDescription('Mesajları onayla temizler.').setDefaultMemberPermissions(PermissionFlagsBits.ManageMessages)
    .addIntegerOption((o) => o.setName('sayi').setDescription('Silinecek mesaj sayısı').setRequired(true).setMinValue(1).setMaxValue(100))
    .addUserOption((o) => o.setName('kullanici').setDescription('Mesajları filtrelenecek kullanıcı'))
];
for (const definition of definitions) definition.setDMPermission(false);

const mappings = {
  sunucu: ['serverinfo', () => []],
  kullanici: ['userinfo', (i) => [i.options.getUser('kullanici')?.id].filter(Boolean)],
  avatar: ['avatar', (i) => [i.options.getUser('kullanici')?.id].filter(Boolean)],
  oyunlar: ['games', () => []],
  tdk: ['tdk', (i) => [i.options.getString('kelime')]],
  webara: ['wsearch', (i) => [i.options.getString('sorgu')]],
  zar: ['roll', (i) => [i.options.getString('zar')].filter(Boolean)],
  ticketsetup: ['ticketsetup', (i) => [
    i.options.getChannel('kategori').id,
    i.options.getChannel('panel_kanali').id,
    i.options.getChannel('log_kanali')?.id,
    i.options.getRole('destek_rolu')?.id
  ].filter(Boolean)],
  yazitura: ['coinflip', () => []],
  sekiztop: ['8ball', (i) => [i.options.getString('soru')]],
  prefix: ['prefix', (i) => [i.options.getString('deger')]],
  cekilis: ['giveaway', (i) => [i.options.getString('sure'), String(i.options.getInteger('kazanan')), i.options.getString('odul')]],
  ticket: ['ticket', (i) => {
    const subcommand = i.options.getSubcommand();
    if (subcommand === 'adlandir') return ['adlandır', i.options.getString('ad')];
    return [subcommand === 'ekle' ? 'ekle' : 'çıkar', i.options.getUser('kullanici').id];
  }],
  unban: ['unban', (i) => [i.options.getString('kullanici_id'), i.options.getString('sebep')].filter(Boolean)],
  cekilisyonet: ['giveawaymanage', (i) => [i.options.getSubcommand(), String(i.options.getInteger('id')), String(i.options.getInteger('kazanan') || '')].filter(Boolean)],
  cases: ['cases', (i) => i.options.getUser('kullanici') ? [i.options.getUser('kullanici').id] : []],
  modconfig: ['modconfig', (i) => i.options.getSubcommand() === 'uyari' ? ['warn', String(i.options.getInteger('esik')), i.options.getString('sure')] : ['appeal', i.options.getChannel('kanal')?.id || 'kapalı']],
  itiraz: ['appeal', (i) => [i.options.getString('metin')]],
  ban: ['ban', (i) => [i.options.getUser('kullanici').id, i.options.getString('sebep'), `--days=${i.options.getInteger('mesaj_sil') || 0}`].filter(Boolean)],
  kick: ['kick', userReasonArgs], unmute: ['unmute', userReasonArgs], warn: ['warn', userReasonArgs],
  mute: ['mute', (i) => [i.options.getUser('kullanici').id, i.options.getString('sure'), i.options.getString('sebep')].filter(Boolean)],
  uyarilar: ['warnings', (i) => [i.options.getUser('kullanici')?.id].filter(Boolean)],
  uyaritemizle: ['clearwarns', (i) => [i.options.getUser('kullanici').id]],
  temizle: ['clear', (i) => [String(i.options.getInteger('sayi')), i.options.getUser('kullanici')?.id].filter(Boolean)],
  kilit: ['lock', (i) => [i.options.getString('islem')]],
  yavasmod: ['slowmode', (i) => [i.options.getString('sure')]]
};

function userReasonArgs(interaction) {
  return [interaction.options.getUser('kullanici').id, interaction.options.getString('sebep')].filter(Boolean);
}

function messageAdapter(interaction) {
  const selectedUser = interaction.options.getUser('kullanici');
  const users = new Collection();
  if (selectedUser) users.set(selectedUser.id, selectedUser);
  let initialResponse = true;
  return {
    id: interaction.id,
    author: interaction.user,
    member: interaction.member,
    guild: interaction.guild,
    channel: interaction.channel,
    client: interaction.client,
    createdTimestamp: interaction.createdTimestamp,
    mentions: { users: { first: () => users.first() } },
    async reply(payload) {
      if (initialResponse) {
        initialResponse = false;
        return interaction.editReply(payload);
      }
      return interaction.followUp(payload);
    },
    async delete() {}
  };
}

async function runSlash(interaction) {
  try {
    if (interaction.commandName === 'yardim') return interaction.reply({ ...help.buildHelp(interaction.client, database.getPrefix(interaction.guildId), 'genel', interaction.user.id), flags: EPHEMERAL });
    if (interaction.commandName === 'kurulum') return interaction.reply({ ...setup.buildSetupPanel(interaction.guild), flags: EPHEMERAL });
    if (interaction.commandName === 'ping') return interaction.reply(ping.buildOutput(interaction.client, interaction.guildId));
    const mapping = mappings[interaction.commandName];
    if (!mapping) return interaction.reply({ content: 'slash komutu eşleştirilemedi.', flags: EPHEMERAL });
    await interaction.deferReply();
    const [commandName, argsBuilder] = mapping;
    const command = interaction.client.commands.get(commandName);
    if (!command) return interaction.editReply('komut yüklenemedi.');
    database.logCommand(interaction.guildId, interaction.user.id, commandName, argsBuilder(interaction).join(' '));
    return await command.execute(messageAdapter(interaction), argsBuilder(interaction), interaction.client);
  } catch (error) {
    console.error(`[HATA] runSlash (${interaction.commandName}):`, error);
    const message = 'Komut çalıştırılırken bir hata oluştu. Yetkileri ve rol sırasını kontrol et.';
    if (interaction.deferred || interaction.replied) {
      await interaction.editReply({ content: message }).catch(() => null);
    } else {
      await interaction.reply({ content: message, flags: EPHEMERAL }).catch(() => null);
    }
  }
}

module.exports = { slashCommands: definitions.map((item) => item.toJSON()), runSlash, messageAdapter };

File: internal/bot/slash.go
package bot

import "github.com/bwmarrin/discordgo"

func textOption(name, desc string, required bool) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionString, Name: name, Description: desc, Required: required}
}
func integerOption(name, desc string, required bool, min, max float64) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionInteger, Name: name, Description: desc, Required: required, MinValue: &min, MaxValue: max}
}
func userOption(required bool) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionUser, Name: "kullanici", Description: "Hedef kullanıcı", Required: required}
}
func channelOption(name, desc string, required bool, types ...discordgo.ChannelType) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionChannel, Name: name, Description: desc, Required: required, ChannelTypes: types}
}
func roleOption(name, desc string) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionRole, Name: name, Description: desc}
}
func subcommand(name, desc string, options ...*discordgo.ApplicationCommandOption) *discordgo.ApplicationCommandOption {
	return &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionSubCommand, Name: name, Description: desc, Options: options}
}
func command(name, desc string, options ...*discordgo.ApplicationCommandOption) *discordgo.ApplicationCommand {
	dm := false
	return &discordgo.ApplicationCommand{Name: name, Description: desc, Options: options, DMPermission: &dm}
}
func permitted(c *discordgo.ApplicationCommand, p int64) *discordgo.ApplicationCommand {
	c.DefaultMemberPermissions = &p
	return c
}

func slashCommands() []*discordgo.ApplicationCommand {
	reason := func() *discordgo.ApplicationCommandOption { return textOption("sebep", "İşlem sebebi", false) }
	manageGuild := int64(discordgo.PermissionManageServer)
	manageChannels := int64(discordgo.PermissionManageChannels)
	manageMessages := int64(discordgo.PermissionManageMessages)
	moderate := int64(discordgo.PermissionModerateMembers)
	ban := int64(discordgo.PermissionBanMembers)
	return []*discordgo.ApplicationCommand{
		command("yardim", "Etkileşimli yardım menüsünü açar."), permitted(command("kurulum", "Gelişmiş sunucu kurulum panelini açar."), manageGuild), command("ping", "Botun bağlantı ve sistem durumunu gösterir."), permitted(command("embed", "Yöneticiler için embed oluşturucu açar."), manageGuild), command("sunucu", "Sunucu bilgilerini gösterir."),
		command("kullanici", "Kullanıcı bilgilerini gösterir.", userOption(false)), command("avatar", "Kullanıcının avatarını gösterir.", userOption(false)), command("oyunlar", "Kanal oyunlarının durumunu gösterir."), command("tdk", "TDK sözlükte ayrıntılı arama yapar.", textOption("kelime", "Aranacak kelime", true)), command("webara", "Web araması yapar.", textOption("sorgu", "Arama sorgusu", true)), command("zar", "Zar atar.", textOption("zar", "Örnek: 2d20", false)), command("yazitura", "Yazı tura atar."), command("sekiztop", "Sihirli küreye soru sorar.", textOption("soru", "Sorun", true)), permitted(command("prefix", "Komut ön ekini değiştirir.", textOption("deger", "Yeni prefix", true)), manageGuild),
		permitted(command("ticketsetup", "Ticket sistemini tek adımda kurar.", channelOption("kategori", "Ticket kategorisi", true, discordgo.ChannelTypeGuildCategory), channelOption("panel_kanali", "Panel kanalı", true, discordgo.ChannelTypeGuildText), channelOption("log_kanali", "Log kanalı", false, discordgo.ChannelTypeGuildText), roleOption("destek_rolu", "Destek rolü")), manageGuild),
		command("ticket", "Açık ticket kanalını yönetir.", subcommand("ekle", "Kullanıcı ekler", userOption(true)), subcommand("cikar", "Kullanıcı çıkarır", userOption(true)), subcommand("adlandir", "Kanalı adlandırır", textOption("ad", "Yeni kanal adı", true))),
		permitted(command("cekilis", "Gelişmiş çekiliş oluşturur.", textOption("sure", "10m, 2h, 3d", true), integerOption("kazanan", "Kazanan sayısı", true, 1, 20), textOption("odul", "Çekiliş ödülü", true)), manageGuild),
		permitted(command("cekilisyonet", "Çekilişi bitirir veya yeniden çeker.", subcommand("bitir", "Erken bitirir", integerOption("id", "Çekiliş ID", true, 1, 999999999)), subcommand("yeniden", "Yeniden çeker", integerOption("id", "Çekiliş ID", true, 1, 999999999), integerOption("kazanan", "Kazanan sayısı", false, 1, 20))), manageGuild),
		command("ban", "Kullanıcıyı onayla yasaklar.", userOption(true), reason(), integerOption("mesaj_sil", "Silinecek geçmiş mesaj günü", false, 0, 7)), command("kick", "Kullanıcıyı onayla sunucudan atar.", userOption(true), reason()), permitted(command("unban", "Yasağı onayla kaldırır.", textOption("kullanici_id", "Yasaklı kullanıcı ID", true), reason()), ban), command("mute", "Kullanıcıyı onayla susturur.", userOption(true), textOption("sure", "10m, 1h, 2d", true), reason()), command("unmute", "Susturmayı onayla kaldırır.", userOption(true), reason()), command("warn", "Kullanıcıya onayla uyarı verir.", userOption(true), reason()), command("uyarilar", "Aktif uyarıları gösterir.", userOption(false)), permitted(command("uyaritemizle", "Aktif uyarıları temizler.", userOption(true)), moderate), permitted(command("cases", "Moderasyon vaka geçmişini gösterir.", userOption(false)), moderate),
		permitted(command("modconfig", "Moderasyon otomasyonunu ayarlar.", subcommand("uyari", "Uyarı cezasını ayarla", integerOption("esik", "Uyarı eşiği", true, 1, 10), textOption("sure", "10m, 1h, 2d", true)), subcommand("itiraz", "İtiraz kanalını ayarla", channelOption("kanal", "İtiraz kanalı", false, discordgo.ChannelTypeGuildText))), manageGuild), command("itiraz", "Yetkili ekibe gizli itiraz gönderir.", textOption("metin", "Ayrıntılı itiraz", true)),
		permitted(command("kilit", "Kanal kilidini yönetir.", &discordgo.ApplicationCommandOption{Type: discordgo.ApplicationCommandOptionString, Name: "islem", Description: "İşlem", Required: true, Choices: []*discordgo.ApplicationCommandOptionChoice{{Name: "Kilitle", Value: "kilitle"}, {Name: "Kilidi Aç", Value: "aç"}}}), manageChannels), permitted(command("yavasmod", "Kanal yavaş modunu ayarlar.", textOption("sure", "0, 10s, 5m, 1h", true)), manageChannels), permitted(command("temizle", "Mesajları temizler.", integerOption("sayi", "Mesaj sayısı", true, 1, 100), userOption(false)), manageMessages),
	}
}
